// udeno@0.3.6/dist/index.cjs
'use strict';

const index = require('./shared/udeno.75b7308b.cjs');
require('destr');
require('fast-glob');
require('defu');
require('node:path');
const promises = require('node:fs/promises');
require('node:process');
require('node:events');
require('node:os');
require('os');
require('path');
require('fs');
require('util');
require('constants');
require('stream');
require('assert');
require('events');
require('node:fs');
require('node:url');
require('node:stream');
require('mlly');
require('consola');
require('node:https');
require('local-pkg');



exports.clean = index.clean;
exports.createReplace = index.createReplace;
exports.createTransformTextFile = index.createTransformTextFile;
exports.defaultNormalize = index.defaultNormalize;
exports.ensureFile = index.ensureFile;
exports.find = index.find;
exports.findDynamicImports = index.findDynamicImports;
exports.findExports = index.findExports;
exports.findStaticImports = index.findStaticImports;
exports.generateVscodeSetting = index.generateVscodeSetting;
exports.getRemoteStdVersion = index.getRemoteStdVersion;
exports.normalizeUrl = index.normalizeUrl;
exports.readTextFile = index.readTextFile;
exports.reduce = index.reduce;
exports.transformReadMe = index.transformReadMe;
exports.udeno = index.udeno;
exports.writeTextFile = promises.writeFile;
